aptos-move/framework/src/natives/function_info.rs

```diff
@@ -0,0 +1,178 @@
+    arguments: &mut VecDeque<Value>,
+) -> SafeNativeResult<(ModuleId, Identifier)> {
+    match arguments.pop_back() {
+        Some(val) => match val.value_as::<StructRef>() {
+            Ok(v) => {
+                let module_address = v
+                    .borrow_field(0)
+                    .and_then(|v| v.value_as::<Reference>())
+                    .and_then(|v| v.read_ref())
+                    .and_then(|v| v.value_as::<AccountAddress>())
+                    .map_err(SafeNativeError::InvariantViolation)?;
+
+                let module_name = identifier_from_ref(
+                    v.borrow_field(1)
+                        .map_err(SafeNativeError::InvariantViolation)?,
+                )?;
+
+                let func_name = identifier_from_ref(
+                    v.borrow_field(2)
+                        .map_err(SafeNativeError::InvariantViolation)?,
+                )?;
+                Ok((ModuleId::new(module_address, module_name), func_name))
+            },
+            Err(e) => Err(SafeNativeError::InvariantViolation(e)),
+        },
+        None => Err(SafeNativeError::InvariantViolation(PartialVMError::new(
+            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+        ))),
+    }
+}
+
+/***************************************************************************************************
+ * native fun check_dispatch_type_compatibility_impl
+ *
+ *   Returns true if the function argument types of rhs is the same as (arguments type of lhs || &FunctionInfo)
+ *   gas cost: base_cost + unit_cost * type_size
+ *
+ **************************************************************************************************/
+fn native_check_dispatch_type_compatibility_impl(
+    context: &mut SafeNativeContext,
+    _ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(arguments.len() == 2);
+
+    // TODO: Figure out the correct gas charging schema here.
+    //
+    // We need to load the modules from lhs and rhs, and cloning the bytes for module id and function name.
+    context.charge(FUNCTION_INFO_CHECK_DISPATCH_TYPE_COMPATIBILITY_IMPL_BASE)?;
+
+    let rhs = {
+        let (module, func) = extract_function_info(&mut arguments)?;
+        context
+            .load_function(&module, &func)
+            .map_err(|_| SafeNativeError::Abort { abort_code: 2 })?
+    };
+    let lhs = {
+        let (module, func) = extract_function_info(&mut arguments)?;
+        context
+            .load_function(&module, &func)
+            .map_err(|_| SafeNativeError::Abort { abort_code: 2 })?
+    };
+
+    if lhs.parameter_types.is_empty() {
+        return Err(SafeNativeError::Abort { abort_code: 2 });
+    }
+
+    Ok(smallvec![Value::bool(
+        rhs.type_parameters == lhs.type_parameters
+            && rhs.return_types == lhs.return_types
+            && lhs.parameter_types[0..lhs.parameter_types.len() - 1] == rhs.parameter_types
```

Comment:
why, what's going on here?

Reply:
https://github.com/aptos-labs/aptos-core/blob/bd8b68c11a14797cb767675b07e1843bfbb0c212/aptos-move/framework/aptos-framework/sources/function_info.move#L41

I have the comments here: we need to check the signature matches other than the last parameter type will always be `&FunctionInfo` so we need to ignore this.

Reply:
It is a little bit confusing to the developers. Would we want to create empty functions w/o last argument as a template?